source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+                                                 TransactionIdStore_t * pTransactionIdStore )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+
+    if( ( pTransactionIdStore == NULL ) || ( maxIdCount > ICE_MAX_STORED_TRANSACTION_ID_COUNT ) || ( maxIdCount < 0 ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        pTransactionIdStore->pTransactionIds = ( uint8_t * ) ( pTransactionIdStore + 1 );
+        pTransactionIdStore->maxTransactionIdsCount = maxIdCount;
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_TransactionIdStoreInsert - Inserts the Transaction in the IceAgent Transaction ID Store. */
+
+static void Ice_TransactionIdStoreInsert( TransactionIdStore_t * pTransactionIdStore,
+                                          uint8_t * pTransactionId )
+{
+    uint8_t * storeLocation = NULL;
+    uint32_t transactionIDCount;
+
+    if( pTransactionIdStore != NULL )
+    {
+        storeLocation = pTransactionIdStore->pTransactionIds + ( ( pTransactionIdStore->nextTransactionIdIndex % pTransactionIdStore->maxTransactionIdsCount ) * STUN_HEADER_TRANSACTION_ID_LENGTH );
+        memcpy( storeLocation,
+                pTransactionId,
+                STUN_HEADER_TRANSACTION_ID_LENGTH );
+
+        pTransactionIdStore->nextTransactionIdIndex = ( pTransactionIdStore->nextTransactionIdIndex + 1 ) % pTransactionIdStore->maxTransactionIdsCount;
+
+        if( pTransactionIdStore->nextTransactionIdIndex == pTransactionIdStore->earliestTransactionIdIndex )
+        {
+            pTransactionIdStore->earliestTransactionIdIndex =
+                ( pTransactionIdStore->earliestTransactionIdIndex + 1 ) % pTransactionIdStore->maxTransactionIdsCount;
+            return;
+        }
+
+        transactionIDCount = ( ( pTransactionIdStore->transactionIdCount + 1 ) > ( pTransactionIdStore->maxTransactionIdsCount ) ) ? pTransactionIdStore->maxTransactionIdsCount : ( pTransactionIdStore->transactionIdCount + 1 );
+
+        pTransactionIdStore->transactionIdCount = transactionIDCount;
+    }
+
+    return;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_TransactionIdStoreHasId - Checks if the transaction is present in the Transaction ID Store. */
+
+static bool Ice_TransactionIdStoreHasId( TransactionIdStore_t * pTransactionIdStore,
+                                         uint8_t * pTransactionId )
+{
+    bool idFound = false;
+    int i, j;
+
+    if( pTransactionIdStore != NULL )
+    {
+        for( i = pTransactionIdStore->earliestTransactionIdIndex, j = 0; j < pTransactionIdStore->maxTransactionIdsCount && !idFound; ++j )
+        {
+            if( memcmp( pTransactionId,
+                        pTransactionIdStore->pTransactionIds + i * STUN_HEADER_TRANSACTION_ID_LENGTH,
```

Comment:
please add braces for correct operator precedence.

Reply:
updated.